File: src-tauri/src/host/lifecycle/inbox.rs
//! The Inbox projection: `inbox_events` and folded threads as cards (#22).
//!
//! Decision #5 fixes the classification and this module does not invent a
//! parallel one. **Still sleeping** is `threads.state = folded` and nothing
//! else — folding writes no card, because the thread row already says it.
//! **Needs you / Done** (and failed, stuck, lost) are `inbox_events` rows, read
//! back with the latest run beside them.
//!
//! What this file adds on top of the raw rows is the *join the host owns*: who
//! the card is from, where the work happened, and what the ledger last said.
//! The renderer could not do that join itself without a second source of truth
//! — it would have to guess a bot's face from a crew list that may not have
//! answered yet, and guess a run's state from the card's own copy.
//!
//! Two things are load-bearing enough to say out loud:
//!
//! **`failed` and `stuck` stay apart.** They are different asks (`resurface.rs`
//! says why) and they arrive here as different `kind`s. Nothing in this module
//! folds one into the other, and `run_state` is what lets a `stuck` card admit
//! the process behind it is still running.
//!
//! **Archived work leaves the Inbox.** `state-machine.md` gives `archived` as
//! "Sidebar hidden / Inbox hidden", and `count_unread_inbox` already refuses to
//! badge it. Listing the card anyway left Archive — one of the two buttons on
//! every resurfaced card — looking like it had done nothing.

use crate::host::protocol::methods::{FoldPolicy, InboxEventView, SleepingThreadView};
use crate::host::store::{BotRow, InboxEventRow, ThreadRow};

use super::state::ThreadState;

/// Whether a card's thread has been closed out. Deleted threads never reach
/// here — the store's own query drops them — so this is about archive.
///
/// A card whose thread row is missing entirely is *not* hidden: it is a row we
/// can no longer explain, and dropping it silently would lose a result. It
/// renders with the deleted state, which is the honest label for it.
pub fn hidden_from_inbox(thread: Option<&ThreadRow>) -> bool {
    thread.is_some_and(|row| super::effective_state(row) == ThreadState::Archived)
}

/// One `inbox_events` row, with everything its card draws.
pub fn event_view(
    row: InboxEventRow,
    thread: Option<&ThreadRow>,
    bot: Option<&BotRow>,
    run_state: Option<String>,
) -> InboxEventView {
    InboxEventView {
        id: row.id,
        thread_id: row.thread_id,
        thread_title: thread.map(|t| t.title.clone()).unwrap_or_default(),
        thread_state: thread
            .map(|t| super::effective_state(t).as_str().to_string())
            .unwrap_or_else(|| ThreadState::Deleted.as_str().to_string()),
        kind: row.kind,
        title: row.title,
        summary: row.summary,
        run_id: row.run_id,
        payload: row
            .payload_json
            .as_deref()
            .and_then(|raw| serde_json::from_str(raw).ok()),
        created_at: row.created_at,
        read_at: row.read_at,
        dismissed_at: row.dismissed_at,
        // The bot is resolved from the *thread*, not from the event: an event
        // has no owner of its own, and a thread that belongs to no bot is a
        // code session, which draws the other avatar.
        bot_id: thread.and_then(|t| t.bot_id.clone()),
        bot_name: bot.map(|b| b.name.clone()),
        bot_color: bot.map(|b| b.color.clone()),
        repo: thread.and_then(|t| t.repo.clone()),
        branch: thread.and_then(|t| t.branch.clone()),
        run_state,
        last_error: thread.and_then(|t| t.last_error.clone()),
        folded_at: thread.and_then(|t| t.folded_at.clone()),
        resurfaced_at: thread.and_then(|t| t.resurfaced_at.clone()),
    }
}

/// One folded thread, as a Still Sleeping row.
pub fn sleeping_view(
    row: &ThreadRow,
    bot: Option<&BotRow>,
    run_state: Option<String>,
    acp_state: &str,
) -> SleepingThreadView {
    SleepingThreadView {
        thread_id: row.id.clone(),
        title: row.title.clone(),
        fold_policy: FoldPolicy::parse(&row.fold_policy),
        folded_at: row.folded_at.clone(),
        run_state,
        acp_state: acp_state.to_string(),
        bot_id: row.bot_id.clone(),
        bot_name: bot.map(|b| b.name.clone()),
        bot_color: bot.map(|b| b.color.clone()),
        repo: row.repo.clone(),
        branch: row.branch.clone(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn thread(state: &str) -> ThreadRow {
        ThreadRow {
            id: "t-1".into(),
            folder_id: Some("f-1".into()),
            bot_id: None,
            harness_id: "claude".into(),
            acp_session_id: None,
            native_session_ref: None,
            cwd: "/tmp".into(),
            runtime_json: "{}".into(),
            title: "Auth migration".into(),
            state: state.into(),
            fold_policy: "default".into(),
            last_stop_reason: None,
            last_error: None,
            preview: None,
            worktree_path: None,
            created_at: "2026-08-20T10:00:00Z".into(),
            updated_at: "2026-08-20T10:00:00Z".into(),
            folded_at: Some("2026-08-20T10:05:00Z".into()),
            resurfaced_at: None,
            archived_at: None,
            deleted_at: None,
            resurfaced_reason: None,
            repo_root: Some("/tmp".into()),
            repo: Some("jabreeflor/jabot".into()),
            forge_host: Some("github.com".into()),
            branch: Some("jabot/t-1".into()),
            host_id: None,
        }
    }

    fn bot() -> BotRow {
        BotRow {
            id: "inboxm".into(),
            name: "Inbox Mgr".into(),
            color: "b-purple".into(),
            instructions: String::new(),
            tools_json: "[]".into(),
            harness_id: "claude".into(),
            is_chief: false,
            template_id: None,
            host_id: None,
            sort_order: 1,
            created_at: "2026-08-20T10:00:00Z".into(),
            updated_at: "2026-08-20T10:00:00Z".into(),
        }
    }

    fn event(kind: &str) -> InboxEventRow {
        InboxEventRow {
            id: format!("e-{kind}"),
            thread_id: "t-1".into(),
            run_id: Some("r-1".into()),
            kind: kind.into(),
            title: "Auth migration failed".into(),
            summary: "the adapter refused".into(),
            payload_json: Some(r#"{"reason":"failed"}"#.into()),
            created_at: "2026-08-20T10:06:00Z".into(),
            read_at: None,
            dismissed_at: None,
        }
    }

    /// Archive is one of the two buttons on a resurfaced card. If the card
    /// survived the click the button would look broken.
    #[test]
    fn archived_work_leaves_the_inbox_and_nothing_else_does() {
        assert!(hidden_from_inbox(Some(&thread("archived"))));
        assert!(!hidden_from_inbox(Some(&thread("resurfaced"))));
        assert!(!hidden_from_inbox(Some(&thread("folded"))));
        assert!(!hidden_from_inbox(Some(&thread("active"))));
        // A card we can no longer explain still gets shown, not swallowed.
        assert!(!hidden_from_inbox(None));
    }

    /// #15 keeps these two reasons apart deliberately; the projection must not
    /// quietly re-merge them on the way to a card.
    #[test]
    fn failed_and_stuck_reach_the_card_as_different_kinds() {
        let failed = event_view(event("failed"), Some(&thread("resurfaced")), None, None);
        let stuck = event_view(
            event("stuck"),
            Some(&thread("resurfaced")),
            None,
            Some("running".into()),
        );

        assert_eq!(failed.kind, "failed");
        assert_eq!(stuck.kind, "stuck");
        // And the difference is legible: a stuck thread's process is still
        // going, which is the fact that makes "wait" a sane answer to it.
        assert_eq!(stuck.run_state.as_deref(), Some("running"));
        assert_eq!(failed.run_state, None);
    }

    #[test]
    fn a_bots_card_carries_the_bot_and_a_code_threads_does_not() {
        let mut owned = thread("resurfaced");
        owned.bot_id = Some("inboxm".into());
        let bot = bot();

        let from_bot = event_view(event("needs_you"), Some(&owned), Some(&bot), None);
        assert_eq!(from_bot.bot_id.as_deref(), Some("inboxm"));
        assert_eq!(from_bot.bot_name.as_deref(), Some("Inbox Mgr"));
        assert_eq!(from_bot.bot_color.as_deref(), Some("b-purple"));

        let from_code = event_view(event("done"), Some(&thread("resurfaced")), None, None);
        assert_eq!(from_code.bot_id, None);
        assert_eq!(from_code.bot_name, None);
        // The code session says where it worked instead.
        assert_eq!(from_code.repo.as_deref(), Some("jabreeflor/jabot"));
        assert_eq!(from_code.branch.as_deref(), Some("jabot/t-1"));
    }

    /// A card whose thread row has gone reports the tombstone rather than
    /// inventing an active thread the user could be sent to.
    #[test]
    fn a_card_without_a_thread_says_so() {
        let orphan = event_view(event("done"), None, None, None);
        assert_eq!(orphan.thread_state, "deleted");
        assert_eq!(orphan.thread_title, "");
        assert_eq!(orphan.bot_id, None);
    }

    #[test]
    fn a_sleeping_row_carries_the_face_asleep_behind_it() {
        let mut owned = thread("folded");
        owned.bot_id = Some("inboxm".into());

        let view = sleeping_view(&owned, Some(&bot()), Some("running".into()), "running");

        assert_eq!(view.thread_id, "t-1");
        assert_eq!(view.bot_name.as_deref(), Some("Inbox Mgr"));
        assert_eq!(view.folded_at.as_deref(), Some("2026-08-20T10:05:00Z"));
        // Folded and still working is the product's whole premise (#5); the row
        // has to be able to say both at once.
        assert_eq!(view.run_state.as_deref(), Some("running"));
        assert_eq!(view.acp_state, "running");
    }
}

File: src/views/inbox.ts
//! The Inbox on real data (#22).
//!
//! Decision #5 makes the Inbox a *projection*: `inbox_events` for what came
//! back, `threads.state = folded` for what is still asleep, and the host owns
//! both. So this hook is the same shape as `folders.ts`, `crew.ts` and
//! `pulls.ts` — ask the host, rename the wire rows into the props the view
//! already takes, and let the fixtures stand only until the first answer.
//!
//! Two things here are not a rename.
//!
//! **An outstanding permission is a card.** The desktop draws an ask inline in
//! the transcript (#20), which is the right place when you are reading the
//! thread — and no place at all when you are not. `permission/pending` is
//! folded into the list by the same [`projectInbox`] the phone uses, so the
//! two devices cannot end up with different ideas of what needs you. Where a
//! thread has both an ask and a `needs_you` card they collapse to one row: the
//! answerable one.
//!
//! **Opening a card is a state change.** `thread/reopen` is what clears a
//! thread's badge (`resurface.md`), puts an archived thread's worktree back
//! (#23), and moves the row out of Still Sleeping and into the sidebar. So the
//! Inbox opens threads through the host rather than by pointing the main pane
//! at an id — and only when the thread is somewhere `reopen` is legal from:
//! the transition table refuses it on a thread that is already active, and a
//! `pr` card's thread very often is.

import { useCallback, useEffect, useMemo, useState } from "react";

import {
  INBOX_EVENT,
  INBOX_RESURFACE,
  PERMISSION_ASK,
  PERMISSION_RESOLVED,
  type HostClient,
  type InboxEventView,
  type NotifyStatusResult,
  type PermissionPendingResult,
  type ThreadOverlayState,
} from "../host";
import type {
  Bot,
  CardSource,
  InboxCard,
  InboxDetail,
  NoticeAction,
} from "../components/types";
// The projection lives under `src/mobile/` because #29 needed it first, and it
// is deliberately device-neutral: its own module docs say two devices
// disagreeing about what needs you would be two products. Importing it is what
// keeps that true — a second copy here is exactly the drift it warns about.
import { projectInbox, type MobileCard as ProjectedCard } from "../mobile/inbox";

/** Open the card's thread. The view raises it through `onOpenThread`. */
export const CARD_REOPEN = "reopen";
/** Close the card's thread out for good. */
export const CARD_ARCHIVE = "archive";
/** `ask:<optionId>` — answer the permission this card *is*, with the agent's
    own option id. Prefixed because a card's other buttons are ours. */
export const ASK_PREFIX = "ask:";

/** The states `thread/reopen` is legal from. Anything else is already open. */
const REOPENABLE: readonly ThreadOverlayState[] = [
  "folded",
  "resurfaced",
  "archived",
];

export interface HostInbox {
  /** `null` until the host answers; `[]` is a real answer and a good day. */
  cards: InboxCard[] | null;
  /** The host's own badge count, not a second classification of these rows. */
  unread: number | null;
  /** Why the last load failed, for the pane that would otherwise look empty. */
  error: string | null;
  loading: boolean;
  reload: () => void;
  /** Open a card's thread, reopening it on the host when it is asleep. */
  open: (threadId: string) => Promise<void>;
  /** Whatever the card's own buttons do. Unknown ids are ignored. */
  act: (cardId: string, actionId: string) => Promise<void>;
  /** What the OS says about banners (#27). `null` until asked, and on any host
      that will not answer — the Inbox is complete without it. */
  notify: NotifyStatusResult | null;
}

/**
 * @param onThreadChanged Re-read the lists this moved a thread between — the
 * sidebar gains a row on reopen and loses one on archive.
 */
export function useInbox(
  client: HostClient | null,
  onThreadChanged?: () => void,
  /** The crew, so a card belonging to a bot can wear that bot's face. `null`
      while it loads, which draws the code mark — see `cardSource`. */
  bots?: readonly Bot[] | null,
): HostInbox {
  const [snapshot, setSnapshot] = useState<Snapshot | null>(null);
  const [notify, setNotify] = useState<NotifyStatusResult | null>(null);
  const [error, setError] = useState<string | null>(null);
  const [loading, setLoading] = useState(false);
  const [generation, setGeneration] = useState(0);

  const reload = useCallback(() => setGeneration((n) => n + 1), []);

  useEffect(() => {
    if (!client) return;
    let cancelled = false;
    setLoading(true);
    // Guarded as a whole, method lookup included: a transport that predates
    // `inbox/list` — an older host, a unit test's stub — should leave the
    // shell on its fixtures rather than take the render down.
    load(client)
      .then((next) => {
        if (cancelled) return;
        setSnapshot(next);
        setError(null);
      })
      .catch((err: unknown) => {
        if (cancelled) return;
        setError(message(err));
      })
      .finally(() => {
        if (!cancelled) setLoading(false);
      });
    return () => {
      cancelled = true;
    };
  }, [client, generation]);

  // Asked once, not on every reload and not inside `load`. An OS notification
  // permission is changed in System Settings, not by anything this app does,
  // so re-asking on every poll would be a round trip per refresh for an answer
  // that does not move. Its failure is swallowed for the same reason
  // `pendingOrNone` swallows its own: the Inbox is complete without it, and
  // every card is written whether or not a banner was ever allowed.
  useEffect(() => {
    if (!client) return;
    if (typeof client.notifyStatus !== "function") return;
    let cancelled = false;
    client
      .notifyStatus()
      .then((status) => {
        if (!cancelled) setNotify(status);
      })
      .catch(() => {
        // Deliberately silent; see above.
      });
    return () => {
      cancelled = true;
    };
  }, [client]);

  // The Inbox is the one view that has to be right while nobody is looking at
  // it: the badge is drawn from it, and the card the user came back for was
  // written by a thread that resurfaced on its own.
  useEffect(() => {
    if (!client) return;
    try {
      return client.onNotification((notification) => {
        if (LIVE.includes(notification.method)) reload();
      });
    } catch {
      // A transport with no notification channel — an older host, a test's
      // stub — still has an Inbox; it just has to be asked for it again.
      return;
    }
  }, [client, reload]);

  const open = useCallback(
    async (threadId: string) => {
      const state = snapshot?.states.get(threadId);
      // Nothing to reopen: the thread is already active, or it is a fixture
      // row this hook has never heard of.
      if (!client || !state || !REOPENABLE.includes(state)) return;
      try {
        await client.reopenThread({ threadId });
        setError(null);
      } catch (err: unknown) {
        setError(message(err));
      } finally {
        reload();
        onThreadChanged?.();
      }
    },
    [client, snapshot, reload, onThreadChanged],
  );

  const act = useCallback(
    async (cardId: string, actionId: string) => {
      const card = snapshot?.cards.find((row) => row.id === cardId);
      const answering = actionId.startsWith(ASK_PREFIX);
      // Nothing to do, and nothing to re-read for: an id this hook does not
      // know is a fixture card's, or a button a later issue added.
      if (!client || !card || (!answering && actionId !== CARD_ARCHIVE)) return;
      try {
        if (answering) {
          const deviceId = client.deviceId;
          if (!deviceId) throw new Error("Not connected to the host yet.");
          // The card *is* the ask, so its id is the request id — the same
          // identity the phone answers with.
          await client.replyPermission({
            requestId: card.id,
            deviceId,
            optionId: actionId.slice(ASK_PREFIX.length),
          });
        } else {
          await client.archiveThread({ threadId: card.threadId });
        }
        setError(null);
      } catch (err: unknown) {
        setError(message(err));
      } finally {
        reload();
        onThreadChanged?.();
      }
    },
    [client, snapshot, reload, onThreadChanged],
  );

  const cards = useMemo(
    () =>
      snapshot ? snapshot.cards.map((card) => cardRow(card, snapshot, bots ?? null)) : null,
    [snapshot, bots],
  );

  return {
    cards,
    unread: snapshot?.unread ?? null,
    error,
    loading,
    reload,
    open,
    act,
    notify,
  };
}

/** The notifications that change what the Inbox says. */
const LIVE: readonly string[] = [
  INBOX_RESURFACE,
  INBOX_EVENT,
  PERMISSION_ASK,
  PERMISSION_RESOLVED,
];

interface Snapshot {
  cards: ProjectedCard[];
  unread: number;
  /** Per thread, so a card knows whether opening it is a reopen. */
  states: Map<string, ThreadOverlayState>;
  /** Per event id, for the journey line a card expands onto. */
  events: Map<string, InboxEventView>;
}

async function load(client: HostClient): Promise<Snapshot> {
  const listed = await client.inbox();
  const pending = await pendingOrNone(client);
  const projection = projectInbox(listed, pending);
  const states = new Map<string, ThreadOverlayState>();
  const events = new Map<string, InboxEventView>();
  for (const event of listed.events) {
    events.set(event.id, event);
    states.set(event.threadId, event.threadState);
  }
  for (const thread of listed.sleeping) states.set(thread.threadId, "folded");
  return {
    // One flat list in the order the view draws them; `InboxView` splits
    // resurfaced from sleeping itself, and filters the tabs off the kind.
    cards: [...projection.needs, ...projection.done, ...projection.sleeping],
    unread: projection.unread,
    states,
    events,
  };
}

/**
 * Asks, or none.
 *
 * A host that will not answer `permission/pending` still has an Inbox, and the
 * events half is the half that resurfaced the thread. Losing the whole list
 * over the answerable extra would be the wrong way round.
 */
async function pendingOrNone(
  client: HostClient,
): Promise<PermissionPendingResult> {
  try {
    return await client.pendingPermissions();
  } catch {
    return { requests: [] };
  }
}

function cardRow(
  card: ProjectedCard,
  snapshot: Snapshot,
  bots: readonly Bot[] | null,
): InboxCard {
  return {
    id: card.id,
    threadId: card.threadId,
    kind: card.kind,
    title: card.title,
    summary: card.summary,
    createdAt: card.at,
    source: cardSource(card.botId, bots),
    detail: detail(card, snapshot),
  };
}

/**
 * The face on an Inbox row.
 *
 * The avatar is the only thing on the row that says *who* this is, and until
 * `inbox/list` carried `botId` every host card wore the generic code mark —
 * including the ones belonging to a named crew member. #22 was right to refuse
 * to invent a bot rather than guess; the fix was to put the id on the wire.
 *
 * Still the code mark in two cases, and both are honest rather than lazy: a
 * thread with no bot really is a code session, and a bot the roster does not
 * (yet) contain cannot be drawn — the crew loads separately, and a card
 * holding only a reference has to draw *something* in between. A face with the
 * wrong name on it would be worse than no face.
 */
function cardSource(botId: string | undefined, bots: readonly Bot[] | null): CardSource {
  if (!botId) return { type: "code" };
  const bot = bots?.find((candidate) => candidate.id === botId);
  if (!bot) return { type: "code" };
  return { type: "bot", name: bot.name, color: bot.color, image: bot.image };
}

/**
 * What a card expands onto.
 *
 * Sleeping rows get nothing: `InboxView` gives them a click that opens the
 * thread instead of a disclosure, because there is nothing to say about work
 * that has not come back yet beyond the fact that it has not.
 */
function detail(
  card: ProjectedCard,
  snapshot: Snapshot,
): InboxDetail | undefined {
  if (card.section === "sleeping") return undefined;
  if (card.ask) {
    return {
      // The command, when the agent sent one: "Run ls" and "Run rm -rf /" are
      // the same title, and this is the line that tells them apart.
      path: card.ask.detail ?? card.title,
      bullets: card.ask.stale
        ? ["The session that asked this is gone — answering records your decision, but the agent will never hear it."]
        : [],
      actions: [
        // The agent's own options, in the agent's order. The host never
        // invents one and neither does this (#20).
        ...card.ask.options.map(
          (option, index): NoticeAction => ({
            id: `${ASK_PREFIX}${option.optionId}`,
            label: option.name,
            primary: index === 0,
          }),
        ),
        { id: CARD_REOPEN, label: "Open thread" },
      ],
    };
  }
  const event = snapshot.events.get(card.id);
  const state = event?.threadState;
  const actions: NoticeAction[] = [
    { id: CARD_REOPEN, label: "Open thread", primary: true },
  ];
  // Archiving an archived thread is an illegal transition, and offering a
  // button whose only outcome is an error message is worse than not offering it.
  if (state !== "archived") {
    actions.push({ id: CARD_ARCHIVE, label: "Archive" });
  }
  return {
    path: event?.threadTitle
      ? `${event.threadTitle} · ${state ?? "thread"}`
      : card.title,
    bullets: [],
    actions,
  };
}

function message(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}
